File: netDxf/IO/DxfRawObjectGraph.cs
// netDxf library. Copyright (c) Daniel Carvajal. Licensed under the MIT License.
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using netDxf.Header;

namespace netDxf.IO
{
    public sealed partial class DxfRawObjectTransaction
    {
        private Change Record(string handle)
        {
            string canonical = DxfObjectText.Handle(handle, false); ulong key = DxfObjectText.Number(canonical);
            if (this.changes.TryGetValue(key, out Change change))
            { if (change.Deleted) throw new KeyNotFoundException("Record was deleted: " + handle); return change; }
            IReadOnlyList<DxfRawHandleOccurrence> definitions = this.store.Index.FindDefinitions(canonical);
            if (definitions.Count != 1 || definitions[0].Record == null) throw new ArgumentException("A unique record identity is required: " + handle);
            DxfRawRecord source = definitions[0].Record;
            return new Change { Source = source, Tags = source.Tags.ToList(), Object = source.SectionName == "OBJECTS" ? this.Get(canonical) : null };
        }
        private static int CommonEnd(IReadOnlyList<DxfTag> tags)
        {
            int depth = 0;
            for (int i = 1; i < tags.Count; ++i)
            {
                DxfTag t = tags[i];
                if (t.Code == 102)
                {
                    string text = (string)t.RawValue;
                    if (text.StartsWith("{", StringComparison.Ordinal)) ++depth;
                    else if (text == "}" && depth > 0) --depth;
                    else throw new InvalidDataException("Malformed common control group.");
                }
                else if (depth == 0 && (t.Code == 100 || t.Code == 101 || t.Code == 1001)) return i;
            }
            if (depth != 0) throw new InvalidDataException("Unterminated common control group.");
            return tags.Count;
        }
        private static Tuple<int, int, string> Extension(IReadOnlyList<DxfTag> tags)
        {
            int end = CommonEnd(tags), depth = 0, start = -1;
            string target = null; Tuple<int, int, string> result = null;
            for (int i = 1; i < end; ++i)
            {
                DxfTag t = tags[i];
                if (t.Code == 102)
                {
                    string text = (string)t.RawValue;
                    if (text.StartsWith("{", StringComparison.Ordinal))
                    {
                        if (depth == 0 && text == "{ACAD_XDICTIONARY") { start = i; target = null; }
                        else if (start >= 0) throw new NotSupportedException("Nested extension dictionary controls are not inferred.");
                        ++depth;
                    }
                    else
                    {
                        --depth;
                        if (depth == 0 && start >= 0)
                        {
                            if (result != null || target == null) throw new InvalidDataException("An extension dictionary requires one group-360 reference.");
                            result = Tuple.Create(start, i + 1, target); start = -1;
                        }
                    }
                }
                else if (start >= 0 && t.Code != 999)
                {
                    if (t.Code != 360 || target != null) throw new NotSupportedException("Unsupported extension dictionary slot.");
                    target = DxfObjectText.Handle((string)t.RawValue, false);
                }
            }
            return result;
        }
        private void ReplaceRecord(string handle, Change original, List<DxfTag> tags)
        {
            if (SameTags(original.Tags, tags)) return;
            if (original.Object != null) { this.Replace(original.Object, tags); return; }
            this.Stage(DxfObjectText.Number(handle), new Change { Source = original.Source, Tags = tags });
        }

        /// <summary>Gets or creates a hard-owned extension dictionary on an existing unique record.</summary>
        /// <remarks>Only the recognized common control group is edited; unrelated entity, reactor and private tags are retained.</remarks>
        public string EnsureExtensionDictionary(string ownerHandle)
        { return this.Apply(() => this.EnsureExtension(ownerHandle)); }
        private string EnsureExtension(string ownerHandle)
        {
            string owner = DxfObjectText.Handle(ownerHandle, false); Change original = this.Record(owner);
            if (original.Source != null && original.Source.SectionName != "OBJECTS" && original.Source.SectionName != "ENTITIES" &&
                original.Source.SectionName != "BLOCKS" && original.Source.SectionName != "TABLES")
                throw new NotSupportedException("Extension dictionaries require an object, table record or entity.");
            Tuple<int, int, string> existing = Extension(original.Tags);
            if (existing != null)
            {
                var dictionary = this.Require<DxfRawDictionary>(existing.Item3);
                if (dictionary.OwnerHandle != owner) throw new InvalidDataException("Extension dictionary owner mismatch.");
                return dictionary.Handle;
            }
            this.EnsureRoot();
            var child = this.AddObject("DICTIONARY", owner, new[] { new DxfTag(100, "AcDbDictionary"), new DxfTag(280, (short)1), new DxfTag(281, (short)1) });
            // EnsureRoot may have changed the very same root record.
            original = this.Record(owner);
            var tags = original.Tags.ToList();
            tags.InsertRange(CommonEnd(tags), new[] { new DxfTag(102, "{ACAD_XDICTIONARY"), new DxfTag(360, child.Handle), new DxfTag(102, "}") });
            this.ReplaceRecord(owner, original, tags); return child.Handle;
        }
        /// <summary>Detaches a common extension dictionary and optionally deletes its exclusively owned known subtree.</summary>
        public bool RemoveExtensionDictionary(string ownerHandle, bool deleteOwnedTree = false)
        {
            return this.Apply(() =>
            {
                string owner = DxfObjectText.Handle(ownerHandle, false); Change original = this.Record(owner);
                var extension = Extension(original.Tags); if (extension == null) return false;
                var dictionary = this.Require<DxfRawDictionary>(extension.Item3);
                if (dictionary.OwnerHandle != owner) throw new InvalidDataException("Extension dictionary owner mismatch.");
                var tags = original.Tags.ToList(); tags.RemoveRange(extension.Item1, extension.Item2 - extension.Item1);
                this.ReplaceRecord(owner, original, tags);
                if (deleteOwnedTree) this.DeleteTree(extension.Item3); return true;
            });
        }

        /// <summary>Creates or replaces the block's ACAD_SORTENTS table and enables the HEADER regeneration sort flag.</summary>
        /// <remarks>The conservative export profile starts at AutoCAD 2004. No geometry or visible rendering is evaluated.</remarks>
        public string SetDrawOrder(string blockRecordHandle, IEnumerable<DxfRawSortOrderEntry> entries, bool enableRegeneration = true)
        {
            return this.Apply(() =>
            {
                if (this.store.Document.Version < DxfVersion.AutoCad2004)
                    throw new NotSupportedException("SORTENTSTABLE authoring requires the conservative AutoCAD 2004+ profile.");
                string block = DxfObjectText.Handle(blockRecordHandle, false); Change blockRecord = this.Record(block);
                if (!Equals(blockRecord.Tags[0].RawValue, "BLOCK_RECORD")) throw new ArgumentException("Draw order requires a BLOCK_RECORD handle.");
                var seen = new HashSet<string>(StringComparer.Ordinal);
                var order = this.Collect(entries, e =>
                {
                    if (e == null) throw new ArgumentException("A redraw entry cannot be null.");
                    if (!seen.Add(e.EntityHandle)) throw new ArgumentException("Duplicate redraw entity handle.");
                    this.Reserve(e.SortHandle);
                    Change entity = this.Record(e.EntityHandle);
                    if (entity.Source == null || (entity.Source.SectionName != "ENTITIES" && entity.Source.SectionName != "BLOCKS"))
                        throw new ArgumentException("A redraw entry must identify a graphic entity.");
                    var owners = this.store.Index.GetOccurrences(entity.Source).Where(o => o.Role == DxfRawHandleRole.Owner).ToArray();
                    if (owners.Length != 1 || owners[0].CanonicalHandle != block) throw new ArgumentException("A redraw entity belongs to another block.");
                });
                string ext = this.EnsureExtension(block); var dictionary = this.Require<DxfRawDictionary>(ext);
                var body = new List<DxfTag> { new DxfTag(100, "AcDbSortentsTable"), new DxfTag(330, block) };
                foreach (var entry in order) { body.Add(new DxfTag(331, entry.EntityHandle)); body.Add(new DxfTag(5, entry.SortHandle)); }
                var link = dictionary.Find("ACAD_SORTENTS"); string result;
                if (link == null)
                {
                    var child = this.AddObject("SORTENTSTABLE", ext, body); this.LinkNew(dictionary, "ACAD_SORTENTS", child); result = child.Handle;
                }
                else
                {
                    var value = this.Require<DxfRawSortentsTable>(link.Handle);
                    if (value.OwnerHandle != ext) throw new InvalidDataException("SORTENTSTABLE owner mismatch.");
                    this.Replace(value, DxfObjectSchema.Wrap(value, body)); result = value.Handle;
                }
                if (enableRegeneration) this.enableSortents = true;
                return result;
            });
        }

        private List<DxfRawStoredObject> CurrentObjects()
        {
            var all = new List<DxfRawStoredObject>();
            foreach (DxfRawStoredObject value in this.store.Objects)
            {
                this.cancellation.ThrowIfCancellationRequested();
                if (value.Handle != null && this.changes.TryGetValue(DxfObjectText.Number(value.Handle), out Change changed))
                { if (!changed.Deleted) all.Add(changed.Object); }
                else all.Add(value);
            }
            foreach (ulong key in this.added)
                if (this.changes.TryGetValue(key, out Change addedObject) && !addedObject.Deleted) all.Add(addedObject.Object);
            return all;
        }
        private List<DxfRawStoredObject> OwnedTree(string handle)
        {
            var all = this.CurrentObjects(); var children = new Dictionary<string, List<DxfRawStoredObject>>(StringComparer.Ordinal);
            foreach (DxfRawStoredObject value in all)
                if (value.OwnerHandle != null && value.Handle != null)
                {
                    if (!children.TryGetValue(value.OwnerHandle, out List<DxfRawStoredObject> list)) children.Add(value.OwnerHandle, list = new List<DxfRawStoredObject>());
                    list.Add(value);
                }
            var queue = new Queue<DxfRawStoredObject>(); var seen = new HashSet<string>(StringComparer.Ordinal);
            queue.Enqueue(this.Get(handle) ?? throw new ArgumentException("Unknown stored object.")); var result = new List<DxfRawStoredObject>();
            while (queue.Count != 0)
            {
                this.cancellation.ThrowIfCancellationRequested(); var current = queue.Dequeue();
                if (!seen.Add(current.Handle)) throw new InvalidDataException("Ownership cycle or duplicate object in subtree.");
                if (current is DxfRawOpaqueStoredObject opaque) throw new NotSupportedException("Owned subtree contains an unsupported schema: " + opaque.Reason);
                result.Add(current);
                if (result.Count > this.store.Options.MaximumChanges) throw new InvalidDataException("Owned subtree exceeds the change budget.");
                if (children.TryGetValue(current.Handle, out List<DxfRawStoredObject> nextChildren)) foreach (var child in nextChildren) queue.Enqueue(child);
            }
            return result;
        }

        /// <summary>Clones a schema-known owned dictionary tree into this same drawing with fresh identities and remapped exposed links.</summary>
        /// <remarks>
        /// Common reactors are not copied. Unknown owned classes and private common controls reject.
        /// Arbitrary handles, sort keys and references hidden in application strings/binary data are not translated.
        /// External links remain external; this is not cross-document import or a private-schema evaluator.
        /// </remarks>
        public string CloneDictionaryTree(string sourceHandle, string parentHandle, string name)
        {
            return this.Apply(() =>
            {
                var source = this.Require<DxfRawDictionary>(sourceHandle); var parent = this.Require<DxfRawDictionary>(parentHandle); this.CheckNewKey(parent, name);
                var tree = this.OwnedTree(source.Handle); var mapping = new Dictionary<string, string>(StringComparer.Ordinal);
                foreach (var value in tree)
                {
                    int end = CommonEnd(value.Tags);
                    foreach (DxfTag tag in value.Tags.Take(end))
                        if (tag.Code == 102 && ((string)tag.RawValue).StartsWith("{", StringComparison.Ordinal) &&
                            !Equals(tag.RawValue, "{ACAD_REACTORS") && !Equals(tag.RawValue, "{ACAD_XDICTIONARY"))
                            throw new NotSupportedException("Private common controls require an application-specific cloning policy.");
                    mapping.Add(value.Handle, this.Allocate());
                }
                foreach (var value in tree)
                {
                    int commonEnd = CommonEnd(value.Tags); var tags = new List<DxfTag>(); int skipDepth = 0;
                    for (int i = 0; i < value.Tags.Count; ++i)
                    {
                        DxfTag tag = value.Tags[i];
                        if (i < commonEnd && tag.Code == 102)
                        {
                            string text = (string)tag.RawValue;
                            if (text == "{ACAD_REACTORS" || skipDepth != 0)
                            {
                                if (text.StartsWith("{", StringComparison.Ordinal)) ++skipDepth;
                                else if (text == "}") --skipDepth;
                                continue;
                            }
                        }
                        if (skipDepth != 0) continue;
                        if (i < commonEnd && tag.Code == 5) tags.Add(new DxfTag(5, mapping[value.Handle]));
                        else if (i < commonEnd && tag.Code == 330 && value.Handle == source.Handle)
                            tags.Add(new DxfTag(330, parent.Handle));
                        else if ((tag.Code >= 330 && tag.Code <= 369 || tag.Code == 1005) &&
                            mapping.TryGetValue(DxfObjectText.Handle((string)tag.RawValue, true), out string replacement))
                            tags.Add(new DxfTag(tag.Code, replacement));
                        else tags.Add(tag);
                    }
                    if (value.OwnerHandle == null)
                        tags.Insert(2, new DxfTag(330, value.Handle == source.Handle ? parent.Handle : mapping[value.OwnerHandle]));
                    DxfRawStoredObject clone = DxfObjectSchema.Read(tags.AsReadOnly(), null, this.store.Options.MaximumPayloadTags);
                    if (clone is DxfRawOpaqueStoredObject opaque) throw new InvalidDataException(opaque.Reason);
                    if (this.store.Objects.Count + this.added.Count >= this.store.Options.MaximumObjects) throw new InvalidDataException("OBJECTS budget exceeded.");
                    ulong key = DxfObjectText.Number(clone.Handle);
                    this.Stage(key, new Change { Tags = tags, Object = clone }); this.added.Add(key);
                }
                var copiedRoot = this.Get(mapping[source.Handle]);
                // The parent might lie in the source tree, but source snapshots were read before editing.
                this.LinkNew(this.Require<DxfRawDictionary>(parent.Handle), name, copiedRoot); return copiedRoot.Handle;
            });
        }

        /// <summary>Deletes an unreferenced, schema-known owned subtree. Incoming external or opaque exposed references reject the operation.</summary>
        /// <remarks>Unlink its parent entry or extension dictionary first, or use the combined removal methods.</remarks>
        public void DeleteOwnedTree(string handle) { this.Apply(() => { this.DeleteTree(handle); return true; }); }
        private void DeleteTree(string handle)
        {
            string canonical = DxfObjectText.Handle(handle, false);
            if (canonical == this.root) throw new InvalidOperationException("The root dictionary cannot be deleted.");
            List<DxfRawStoredObject> tree = this.OwnedTree(canonical);
            var deleted = new HashSet<ulong>(tree.Select(o => DxfObjectText.Number(o.Handle)));
            DxfRawDocument current = this.Build(false);
            DxfRawHandleIndex index = DxfRawHandleIndex.Create(current, cancellationToken: this.cancellation);
            var sourceIds = index.Occurrences.Where(o => o.Role == DxfRawHandleRole.Identity && o.Record != null)
                .ToDictionary(o => o.Record, o => o.NumericHandle);
            foreach (DxfRawHandleOccurrence item in index.Occurrences)
            {
                this.cancellation.ThrowIfCancellationRequested();
                if (!deleted.Contains(item.NumericHandle) || item.Role == DxfRawHandleRole.Identity ||
                    item.Role == DxfRawHandleRole.HeaderSeed || item.Role == DxfRawHandleRole.Arbitrary) continue;
                if (item.Record != null && sourceIds.TryGetValue(item.Record, out ulong source) && deleted.Contains(source)) continue;
                throw new InvalidOperationException("Deleting the subtree would invalidate an external or opaque exposed reference to " + item.Handle + ".");
            }
            foreach (var value in tree)
                this.Stage(DxfObjectText.Number(value.Handle), new Change { Source = value.SourceRecord, Deleted = true, Object = value });
        }
    }
}

File: javascript/netDxf/IO/DxfRawObjectGraph.js
// netDxf library. Copyright (c) Daniel Carvajal. Licensed under the MIT License.
import { DxfObjectText as Text, DxfObjectSchema as Schema } from './DxfRawObjectStore.js';
import * as M from './DxfRawObjectModel.js';
import { DxfRawHandleIndex } from './DxfRawHandleIndex.js';
import { DxfRawHandleRole as R } from './DxfRawHandleModel.js';
import { DxfTag } from './DxfTag.js';
import { Get, Require, SameTags, Stage, Replace, EnsureRoot, AddObject, Collect, Reserve,
  LinkNew, Allocate, CheckNewKey } from './DxfRawObjectTransaction.js';
import { Build } from './DxfRawObjectCommit.js';
import * as E from '../../runtime/Errors.js';
import { ReadOnlyList } from '../../runtime/Collections.js';
const T = (code, value) => new DxfTag(code, value);

export function Record(s, handle) {
  const canonical = Text.Handle(handle, false), key = Text.Number(canonical), change = s.changes.get(key);
  if (change) {
    if (change.Deleted) throw new E.KeyNotFoundException('Record was deleted: ' + handle);
    return change;
  }
  const definitions = s.index.FindDefinitions(canonical);
  if (definitions.length !== 1 || definitions[0].Record === null) throw new E.ArgumentException('A unique record identity is required: ' + handle);
  const source = definitions[0].Record;
  return { Source: source, Tags: Array.from(source.Tags), Object: source.SectionName === 'OBJECTS' ? Get(s, canonical) : null };
}
export function CommonEnd(tags) {
  let depth = 0;
  for (let i = 1; i < tags.length; i++) {
    const t = tags[i];
    if (t.Code === 102) {
      const text = t.Value;
      if (text.startsWith('{')) depth++;
      else if (text === '}' && depth > 0) depth--;
      else throw new E.InvalidDataException('Malformed common control group.');
    } else if (depth === 0 && [100, 101, 1001].includes(t.Code)) return i;
  }
  if (depth !== 0) throw new E.InvalidDataException('Unterminated common control group.');
  return tags.length;
}
export function Extension(tags) {
  const end = CommonEnd(tags); let depth = 0, start = -1, target = null, result = null;
  for (let i = 1; i < end; i++) {
    const t = tags[i];
    if (t.Code === 102) {
      const text = t.Value;
      if (text.startsWith('{')) {
        if (depth === 0 && text === '{ACAD_XDICTIONARY') { start = i; target = null; }
        else if (start >= 0) throw new E.NotSupportedException('Nested extension dictionary controls are not inferred.');
        depth++;
      } else {
        depth--;
        if (depth === 0 && start >= 0) {
          if (result !== null || target === null) throw new E.InvalidDataException('An extension dictionary requires one group-360 reference.');
          result = [start, i + 1, target]; start = -1;
        }
      }
    } else if (start >= 0 && t.Code !== 999) {
      if (t.Code !== 360 || target !== null) throw new E.NotSupportedException('Unsupported extension dictionary slot.');
      target = Text.Handle(t.Value, false);
    }
  }
  return result;
}
export function ReplaceRecord(s, handle, original, tags) {
  if (SameTags(original.Tags, tags)) return;
  if (original.Object !== null) { Replace(s, original.Object, tags); return; }
  Stage(s, Text.Number(handle), { Source: original.Source, Tags: tags, Object: null });
}
export function EnsureExtension(s, ownerHandle) {
  const owner = Text.Handle(ownerHandle, false); let original = Record(s, owner);
  if (original.Source !== null && !['OBJECTS', 'ENTITIES', 'BLOCKS', 'TABLES'].includes(original.Source.SectionName))
    throw new E.NotSupportedException('Extension dictionaries require an object, table record or entity.');
  const existing = Extension(original.Tags);
  if (existing !== null) {
    const dictionary = Require(s, M.DxfRawDictionary, existing[2]);
    if (dictionary.OwnerHandle !== owner) throw new E.InvalidDataException('Extension dictionary owner mismatch.');
    return dictionary.Handle;
  }
  EnsureRoot(s);
  const child = AddObject(s, 'DICTIONARY', owner, [T(100, 'AcDbDictionary'), T(280, 1), T(281, 1)]);
  original = Record(s, owner);
  const tags = Array.from(original.Tags);
  tags.splice(CommonEnd(tags), 0, T(102, '{ACAD_XDICTIONARY'), T(360, child.Handle), T(102, '}'));
  ReplaceRecord(s, owner, original, tags); return child.Handle;
}
export function RemoveExtensionDictionary(s, ownerHandle, deleteOwnedTree) {
  const owner = Text.Handle(ownerHandle, false), original = Record(s, owner), extension = Extension(original.Tags);
  if (extension === null) return false;
  const dictionary = Require(s, M.DxfRawDictionary, extension[2]);
  if (dictionary.OwnerHandle !== owner) throw new E.InvalidDataException('Extension dictionary owner mismatch.');
  const tags = Array.from(original.Tags); tags.splice(extension[0], extension[1] - extension[0]);
  ReplaceRecord(s, owner, original, tags);
  if (deleteOwnedTree) DeleteTree(s, extension[2]); return true;
}
export function SetDrawOrder(s, blockRecordHandle, entries, enableRegeneration) {
  if (s.store.Document.Version < 14) throw new E.NotSupportedException('SORTENTSTABLE authoring requires the conservative AutoCAD 2004+ profile.');
  const block = Text.Handle(blockRecordHandle, false), blockRecord = Record(s, block);
  if (blockRecord.Tags[0].Value !== 'BLOCK_RECORD') throw new E.ArgumentException('Draw order requires a BLOCK_RECORD handle.');
  const seen = new Set();
  const order = Collect(s, entries, e => {
    if (!(e instanceof M.DxfRawSortOrderEntry)) throw new E.ArgumentException('A redraw entry cannot be null.');
    if (seen.has(e.EntityHandle)) throw new E.ArgumentException('Duplicate redraw entity handle.');
    seen.add(e.EntityHandle); Reserve(s, e.SortHandle);
    const entity = Record(s, e.EntityHandle);
    if (entity.Source === null || !['ENTITIES', 'BLOCKS'].includes(entity.Source.SectionName))
      throw new E.ArgumentException('A redraw entry must identify a graphic entity.');
    const owners = s.index.GetOccurrences(entity.Source).filter(o => o.Role === R.Owner);
    if (owners.length !== 1 || owners[0].CanonicalHandle !== block) throw new E.ArgumentException('A redraw entity belongs to another block.');
  });
  const ext = EnsureExtension(s, block), dictionary = Require(s, M.DxfRawDictionary, ext);
  const body = [T(100, 'AcDbSortentsTable'), T(330, block)];
  for (const entry of order) body.push(T(331, entry.EntityHandle), T(5, entry.SortHandle));
  const link = dictionary.Find('ACAD_SORTENTS'); let result;
  if (link === null) {
    const child = AddObject(s, 'SORTENTSTABLE', ext, body); LinkNew(s, dictionary, 'ACAD_SORTENTS', child); result = child.Handle;
  } else {
    const value = Require(s, M.DxfRawSortentsTable, link.Handle);
    if (value.OwnerHandle !== ext) throw new E.InvalidDataException('SORTENTSTABLE owner mismatch.');
    Replace(s, value, Schema.Wrap(value, body)); result = value.Handle;
  }
  if (enableRegeneration) s.enableSortents = true;
  return result;
}
export function CurrentObjects(s) {
  const all = [];
  for (const value of s.store.Objects) {
    E.ThrowIfCancellationRequested(s.cancellation);
    const changed = value.Handle !== null ? s.changes.get(Text.Number(value.Handle)) : null;
    if (changed) { if (!changed.Deleted) all.push(changed.Object); } else all.push(value);
  }
  for (const key of s.added) {
    const value = s.changes.get(key);
    if (value && !value.Deleted) all.push(value.Object);
  }
  return all;
}
export function OwnedTree(s, handle) {
  const children = new Map();
  for (const value of CurrentObjects(s)) if (value.OwnerHandle !== null && value.Handle !== null) {
    let list = children.get(value.OwnerHandle);
    if (!list) children.set(value.OwnerHandle, list = []);
    list.push(value);
  }
  const first = Get(s, handle);
  if (!first) throw new E.ArgumentException('Unknown stored object.');
  const queue = [first], seen = new Set(), result = [];
  for (let at = 0; at < queue.length; at++) {
    E.ThrowIfCancellationRequested(s.cancellation); const current = queue[at];
    if (seen.has(current.Handle)) throw new E.InvalidDataException('Ownership cycle or duplicate object in subtree.');
    seen.add(current.Handle);
    if (current instanceof M.DxfRawOpaqueStoredObject) throw new E.NotSupportedException('Owned subtree contains an unsupported schema: ' + current.Reason);
    result.push(current);
    if (result.length > s.options.MaximumChanges) throw new E.InvalidDataException('Owned subtree exceeds the change budget.');
    for (const child of children.get(current.Handle) ?? []) queue.push(child);
  }
  return result;
}
export function CloneDictionaryTree(s, sourceHandle, parentHandle, name) {
  const source = Require(s, M.DxfRawDictionary, sourceHandle), parent = Require(s, M.DxfRawDictionary, parentHandle);
  CheckNewKey(parent, name);
  const tree = OwnedTree(s, source.Handle), mapping = new Map();
  for (const value of tree) {
    const end = CommonEnd(value.Tags);
    for (const tag of value.Tags.slice(0, end))
      if (tag.Code === 102 && tag.Value.startsWith('{') && !['{ACAD_REACTORS', '{ACAD_XDICTIONARY'].includes(tag.Value))
        throw new E.NotSupportedException('Private common controls require an application-specific cloning policy.');
    mapping.set(value.Handle, Allocate(s));
  }
  for (const value of tree) {
    const commonEnd = CommonEnd(value.Tags), tags = []; let skipDepth = 0;
    for (let i = 0; i < value.Tags.length; i++) {
      const tag = value.Tags[i];
      if (i < commonEnd && tag.Code === 102) {
        const text = tag.Value;
        if (text === '{ACAD_REACTORS' || skipDepth !== 0) {
          if (text.startsWith('{')) skipDepth++; else if (text === '}') skipDepth--;
          continue;
        }
      }
      if (skipDepth !== 0) continue;
      if (i < commonEnd && tag.Code === 5) tags.push(T(5, mapping.get(value.Handle)));
      else if (i < commonEnd && tag.Code === 330 && value.Handle === source.Handle) tags.push(T(330, parent.Handle));
      else if ((tag.Code >= 330 && tag.Code <= 369 || tag.Code === 1005) && mapping.has(Text.Handle(tag.Value, true)))
        tags.push(T(tag.Code, mapping.get(Text.Handle(tag.Value, true))));
      else tags.push(tag);
    }
    if (value.OwnerHandle === null) tags.splice(2, 0, T(330, value.Handle === source.Handle ? parent.Handle : mapping.get(value.OwnerHandle)));
    const clone = Schema.Read(ReadOnlyList(tags), null, s.options.MaximumPayloadTags);
    if (clone instanceof M.DxfRawOpaqueStoredObject) throw new E.InvalidDataException(clone.Reason);
    if (s.store.Objects.length + s.added.length >= s.options.MaximumObjects) throw new E.InvalidDataException('OBJECTS budget exceeded.');
    const key = Text.Number(clone.Handle);
    Stage(s, key, { Source: null, Tags: tags, Object: clone, Deleted: false }); s.added.push(key);
  }
  const copiedRoot = Get(s, mapping.get(source.Handle));
  LinkNew(s, Require(s, M.DxfRawDictionary, parent.Handle), name, copiedRoot); return copiedRoot.Handle;
}
export function DeleteTree(s, handle) {
  const canonical = Text.Handle(handle, false);
  if (canonical === s.root) throw new E.InvalidOperationException('The root dictionary cannot be deleted.');
  const tree = OwnedTree(s, canonical), deleted = new Set(tree.map(o => Text.Number(o.Handle)));
  const current = Build(s, false), index = DxfRawHandleIndex.Create(current, null, s.cancellation);
  const sourceIds = new Map(index.Occurrences.filter(o => o.Role === R.Identity && o.Record !== null).map(o => [o.Record, o.NumericHandle]));
  for (const item of index.Occurrences) {
    E.ThrowIfCancellationRequested(s.cancellation);
    if (!deleted.has(item.NumericHandle) || [R.Identity, R.HeaderSeed, R.Arbitrary].includes(item.Role)) continue;
    if (item.Record !== null && sourceIds.has(item.Record) && deleted.has(sourceIds.get(item.Record))) continue;
    throw new E.InvalidOperationException('Deleting the subtree would invalidate an external or opaque exposed reference to ' + item.Handle + '.');
  }
  for (const value of tree) Stage(s, Text.Number(value.Handle), { Source: value.SourceRecord, Deleted: true, Object: value });
}
